src/gazebo_vision_plugin.cpp

```diff
@@ -113,21 +129,214 @@ void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
   _nh = transport::NodePtr(new transport::Node());
   _nh->Init(_namespace);
 
-  // Listen to the update event. This event is broadcast every simulation iteration.
-  _updateConnection = event::Events::ConnectWorldUpdateBegin(
-      boost::bind(&VisionPlugin::OnUpdate, this, _1));
+  if(_enable_ros_odom)
+  {
+#if BUILD_ROS_INTERFACE == 1
+  // ROS Topic subscriber
+  // Initialize ROS, if it has not already bee initialized.
+  if (!ros::isInitialized())  {
+    int argc = 0;
+    char **argv = NULL;
+    ros::init(argc, argv, "gazebo_ros_sub", ros::init_options::NoSigintHandler);
+  }
+
+  // Create our ROS node. This acts in a similar manner to the Gazebo node
+  this->_ros_node.reset(new ros::NodeHandle("gazebo_client"));
+
+  // Create a named topic, and subscribe to it.
+  ros::SubscribeOptions so = ros::SubscribeOptions::create<geometry_msgs::PoseWithCovarianceStamped>(_ros_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCovCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
+  this->_ros_sub = this->_ros_node->subscribe(so);
+
+  this->_ros_queue_thread = std::thread(std::bind(&VisionPlugin::queueThread, this));
+#endif
+  } else {
+    // Listen to the update event. This event is broadcast every simulation iteration.
+    _updateConnection = event::Events::ConnectWorldUpdateBegin(
+        boost::bind(&VisionPlugin::OnUpdate, this, _1));
+  }
 
   _pub_odom = _nh->Advertise<nav_msgs::msgs::Odometry>("~/" + _model->GetName() + "/vision_odom", 10);
 }
 
-void VisionPlugin::OnUpdate(const common::UpdateInfo&)
-{
+double VisionPlugin::calcTimeStep() {
 #if GAZEBO_MAJOR_VERSION >= 9
-  common::Time current_time = _world->SimTime();
+  _current_time = _world->SimTime();
 #else
-  common::Time current_time = _world->GetSimTime();
+  _current_time = _world->GetSimTime();
 #endif
-  double dt = (current_time - _last_pub_time).Double();
+  return (_current_time - _last_pub_time).Double();
+}
+
+#if BUILD_ROS_INTERFACE == 1
+void VisionPlugin::rosPoseCovCallBack(const geometry_msgs::PoseWithCovarianceStampedConstPtr& odomMsg)
+{
+  double dt = calcTimeStep();
+
+  if (dt > 1.0 / _pub_rate) {
```

Comment:
You can put all the bellow code on a specific function:
```c++
void VisionPlugin::handle_vision_data(geometry_msgs::Vector3 &position, geometry_msgs::Quaternion &orientation, geometry_msgs::Vector3 &lin_vel, geometry_msgs::Vector3 &ang_vel, float64[36] &pose_covariance, float64[36] &twist_covariance) {
    // handle all the parsing to gazebo msgs here
    (...)
}

void VisionPlugin::rosPoseCovCallBack(const geometry_msgs::PoseWithCovarianceStampedConstPtr& poseCovMsg)
{
    geometry_msgs::Vector3 lin_vel(0,0,0);
    geometry_msgs::Vector3 ang_vel(0,0,0);
    float64[36] twist_covariance = { }; // all values are zero
    handle_vision_data(poseCov->pose.pose.position, poseCov->pose.pose.orientation, lin_vel, ang_vel, poseCov->pose.covariance, twist_covariance);
}

// then the same for the other Callbacks
```

Reply:
Done.